app/models/calculator/weight.rb

```diff
@@ -13,8 +13,25 @@ def self.description
 
     def compute(object)
       line_items = line_items_for object
-      total_weight = line_items.sum { |li| ((li.variant.andand.weight || 0) * li.quantity) }
-      total_weight * preferred_per_kg
+      total_weight(line_items) * preferred_per_kg
+    end
+
+    private
+
+    def total_weight(line_items)
+      line_items.sum do |line_item|
+        line_item_weight(line_item)
+      end
+    end
+
+    def line_item_weight(line_item)
+      return 0 if line_item.variant.product.andand.variant_unit != 'weight'
```

Comment:
hi @kristinalim from slack you said:
"I think it's not valid to consider 0 weight for items that are not using weight as the unit type."

can you add more details? maybe I can investigate further.

Reply:
@luisramos0 The use case I was thinking of is this: You have a 1L variant, which may or may not be 1kg in weight. The shop supports weight-based shipping, so they set the "Weight" in the variant form to 1.2kg.

With this guard clause, this line item would have 0 weight when calculating shipping, which should not be the case.

This is what I see the calculation would look like:

```
actual_quantity_of_variant = 1.0 * final_units_per_variant / total_units_per_variant
actual_kg_per_variant = actual_quantity_of_variant * kg_per_variant
```

Reply:
sorry @kristinalim I am not sure I follow.
can you clarify what these values are? final_units_per_variant and total_units_per_variant
are you not repeating what's already done in line_item.calculate_final_weight_volume and line_item.unit_value?

I understand that we need to calculate weight even if line_item.variant.product.andand.variant_unit != 'weight'

So, I think that if variant_unit is volume or items but final_weight_volume is not present, we can still just use 
line_item.variant.andand.weight.
so, the only condition to change would be when variant_unit != weight and final_weight_volume.present?

something like this:
```suggestion
      if line_item.final_weight_volume.present?
            if line_item.variant.product.andand.variant_unit == 'weight'
                  # Divided by 1000 because grams is the base weight unit and the calculator price is per_kg
                  line_item.final_weight_volume / 1000
            else
                  line_item.final_weight_volume
            end
      else
        (line_item.variant.andand.weight || 0) * line_item.quantity
      end
```

what do you think?

Reply:
Sorry @luisramos0 if there is something I do not understand correctly or I missed something.

Let us use these examples for a 3 L variant with 2.5 kg weight:

* Variant `unit_value`: 3 (L)
* Product `variant_unit`: volume
* Product `variant_unit_scale`: 1.0
* Variant weight: 2.5 (kg)

And weight-based shipping costing $6 per kg.

If I order 7 units of the variant, the shipping fee should be $105:

```
7 pc * (2.5 kg / pc) * ($6 / kg)
# => 105
```

This order sets "Final Weight/Volume" to 21. If the admin changes this value to 22 (as in 22 L), the code above would change recalculate the fee to $132:

```
22 L * ($6 / kg)
```

When it should be $110:

```
22 L * (2.5 kg / 3 L) * ($6 / kg)
```

I will post another comment with suggested code change.

Reply:
I wanted to verify the logic, and ended up with a full commit. :joy:

@luisramos0 Please take a look here: https://github.com/kristinalim/openfoodnetwork/commit/e1fce8304d1c4cc5da1c5708e0a79c8e3bc449d4 I added specs testing how `final_weight_volume` is handled for "weight", "volume", and "items", and have updated the calculator as well. Let me know if that makes sense.

Because of the division performed for "volume" and "items" variants, I also added `.round(3)` to the fee calculation: https://github.com/kristinalim/openfoodnetwork/commit/0d6ba90ea1b5084b9efc087dc851833826354034